### PR TITLE
Add vector legalization pass and fix some issues

### PR DESCRIPTION
**Pull request by giuseros**
- Adding a legalization pass that works across for-loops
- Change the scheduling pass to work with the legalized version
- Add the possibility to interleave MLIR inside the micro-kernel
- Fixed the Alp runtime to include functions we are using to print time and the PID of the program
- Fixed cmake infra

### CODE REVIEW

File: experimental/alp/lib/Transforms/extract_kernel_pass.cpp [hunk 1]
```diff
@@ -53,8 +60,12 @@ void extract_function(StringRef func_name, Block *block, ModuleOp parentModule,
   // entry block are constants. For all other variables, we will add them as
   // inputs to the function
   block->walk([&](Operation *inst) {
-    for (Value val : inst->getOperands()) {
-      if (dom_info.properlyDominates(val, parent_op)) {
+    for (Value val: inst->getOperands()){
+      if (dom_info.properlyDominates(val, parent_op)){
+        if(func_name=="kernel_tail"){
+        llvm::errs()<<"dominates?\n";
```
Comment (nicolasvasilache) [inline experimental/alp/lib/Transforms/extract_kernel_pass.cpp:66]: debug spew ?

File: experimental/alp/lib/Transforms/extract_kernel_pass.cpp [hunk 2]
```diff
@@ -183,4 +194,149 @@ struct ExtractKernelPass : public ExtractKernelPassBase<ExtractKernelPass> {
...
+  //std::set<int64_t> consts;
+  llvm::SmallVector<Value, 1> vals;
+  llvm::SmallVector<arith::ConstantOp, 1> consts;
+
+  // Walk the block and find out all the variables that were defined outside 
+  // this block and are used inside the block (i.e., all the variables x that
+  // properly dominate the block). The only things we will redefine inside the
+  // entry block are constants. For all other variables, we will add them as 
+  // inputs to the function
+  block->walk([&](Operation *inst) {
```
Comment (nicolasvasilache) [inline experimental/alp/lib/Transforms/extract_kernel_pass.cpp:222]: I've seen this implemented in various places as:
```
bool isDefinedOutsideOfX(Value value) {
  return !region().isAncestor(value.getParentRegion());
}
```

File: experimental/alp/lib/Transforms/legalize_vector.cpp [hunk 3]
```diff
@@ -0,0 +1,211 @@
...
+              extractOp.getResult().replaceAllUsesWith(newForOp.getResults().back());
+              rewriter.eraseOp(extractOp);
+          }
+      }
+    }
+    return success();
+  }
+};
+
+struct LegalizePass : public LegalizePassBase<LegalizePass>{
```
Comment (nicolasvasilache) [inline experimental/alp/lib/Transforms/legalize_vector.cpp:123]: I am not too sure what this pass does as there are not tests but I'd say it looks a bit fishy and at the intersection of other things that should exist in core.
Maybe if you describe the objective here I can provide some links?

File: experimental/alp/lib/Transforms/modulo_scheduling_pass.cpp [hunk 4]
```diff
@@ -22,99 +22,146 @@
...
-        op->setAttr("unrolled_iteration", b.getUI32IntegerAttr(i));
-      };
+namespace{
+  enum StageType{
+    Compute,
+    Load
+  };
 
-      (void)loopUnrollByFactor(loop, 2, annotateFn);
+  struct ModuloSchedulingPass : public ModuloSchedulingPassBase<ModuloSchedulingPass>
```
Comment (nicolasvasilache) [inline experimental/alp/lib/Transforms/modulo_scheduling_pass.cpp:31]: note that there is a loop pipelining pass that does similar things, could it be adapted to your uses?

**Review by nicolasvasilache — APPROVED** [review]

File: experimental/alp/lib/Transforms/modulo_scheduling_pass.cpp (hunk 4 above)
Comment (giuseros) [reply experimental/alp/lib/Transforms/modulo_scheduling_pass.cpp:31]: If you refer to `ForLoopPipelining`, yes, it is exactly what I am using !

File: experimental/alp/lib/Transforms/legalize_vector.cpp (hunk 3 above)
Comment (giuseros) [reply experimental/alp/lib/Transforms/legalize_vector.cpp:123]: We discussed this here: https://llvm.discourse.group/t/outer-product-spilling/4405/7. I am basically chunking writes, reads and contracts in chunks of 4. As you correctly pointed out [here](https://llvm.discourse.group/t/outer-product-spilling/4405/3),  `scf.for + iter_args throw a wrench into the unrolling patterns`. I added the code to remove the `wrench`

File: experimental/alp/lib/Transforms/extract_kernel_pass.cpp (hunk 2 above)
Comment (giuseros) [reply experimental/alp/lib/Transforms/extract_kernel_pass.cpp:222]: Thanks Nicolas! Can I ask more info? What does this replace? I still need to walk every operation of the block, right?

**Comment by giuseros** [comment]
@nicolasvasilache a friendly ping! I have another small set of transformations after this PR which got our performance to 94% of the library (finally).

**Comment by nicolasvasilache** [comment]
@giuseros the PR is accepted, I won't be able to give you much deeper insights on the passes themselves at this time so please just merge at your convenience.
Once you have everything in and we can repro and dump some assembly we can iterate on generalizing some of the transforms; it is possible/likely that some of the stuff you have is also needed to bridge a perf gap I currently see on CPU.

**Comment by nicolasvasilache** [comment]
> I have another small set of transformations after this PR which got our performance to 94% of the library (finally).

This sounds awesome!

**Comment by giuseros** [comment]
Ah, so the problem is that I cannot merge it, as apparently I am not authorized to merge things in :) Sorry for the misunderstanding, I thought you were waiting to merge because wanted to add more replies

**Comment by nicolasvasilache** [comment]
my apologies, it is still the subdir permission thing .. didn't get a cycle to look at that

File: experimental/alp/lib/Transforms/extract_kernel_pass.cpp [hunk 5]
```diff
@@ -183,4 +190,149 @@ struct ExtractKernelPass : public ExtractKernelPassBase<ExtractKernelPass> {
...
+  // Build the dominance tree of the parent op of the block
+  Region *region = block->getParent();
+  Operation *parent_op = region->getParentOp();
+  auto dom_info = mlir::DominanceInfo(parent_op);
+
+  //std::set<int64_t> consts;
+  llvm::SmallVector<Value, 1> vals;
+  llvm::SmallVector<arith::ConstantOp, 1> consts;
+
+  // Walk the block and find out all the variables that were defined outside
```
Comment (nicolasvasilache) [inline experimental/alp/lib/Transforms/extract_kernel_pass.cpp:213]: @giuseros side note, did you see this older commit: https://reviews.llvm.org/D85449 ?

In particular the "defined outside" thing is a common core function.

**Pull request merged**